extension/storage/filestorage/client_test.go

```diff
@@ -209,19 +215,141 @@ func TestNewClientErrorsOnInvalidBucket(t *testing.T) {
+	t.Cleanup(func() {
+		require.NoError(t, client.Close(context.TODO()))
+	})
+
+	// 1. Fill up the database
+	position := int64(0)
+	ctx := context.Background()
+
+	entrySize := int64(1048576)
+
+	for ; position < 5; position++ {
+		batchWrite := []storage.Operation{
+			storage.SetOperation(fmt.Sprintf("foo-%d", position), make([]byte, entrySize)),
+			storage.SetOperation(fmt.Sprintf("bar-%d", position), []byte("testValueBar")),
+		}
+		err = client.Batch(ctx, batchWrite...)
+		require.NoError(t, err)
+	}
+
+	require.Eventually(t,
+		func() bool {
+			totalSize, realSize, dbErr := client.getDbSize()
+			require.NoError(t, dbErr)
+			return totalSize > position*entrySize && realSize > position*entrySize
+		},
+		10*time.Second, 5*time.Millisecond, "database allocated space for data",
+	)
+
+	// 2. Remove the large entries
+	for i := 0; i < int(position); i++ {
+		err = client.Batch(ctx, storage.DeleteOperation(fmt.Sprintf("foo-%d", i)))
+		require.NoError(t, err)
+	}
+
+	require.Eventually(t,
+		func() bool {
+			totalSize, realSize, dbErr := client.getDbSize()
+			require.NoError(t, dbErr)
+			return totalSize < entrySize && realSize < entrySize
+		},
+		10*time.Second, 5*time.Millisecond, "database cleaned up not used space",
+	)
+}
+
+func TestClientConcurrentCompaction(t *testing.T) {
+	tempDir := t.TempDir()
+	dbFile := filepath.Join(tempDir, "my_db")
+
+	checkInterval := time.Millisecond
+
+	client, err := newClient(zap.NewNop(), dbFile, time.Second, &CompactionConfig{
+		OnRebound:                  true,
+		CheckInterval:              checkInterval,
+		ReboundNeededThresholdMiB:  1,
+		ReboundTriggerThresholdMiB: 5,
+	})
+	require.NoError(t, err)
+
+	t.Cleanup(func() {
+		require.NoError(t, client.Close(context.TODO()))
+	})
+
+	var wg sync.WaitGroup
+	repeats := 5
+	ctx := context.Background()
+
+	clientOperationsThread := func(id int) {
+		for i := 0; i < repeats; i++ {
+			batchWrite := []storage.Operation{
+				storage.SetOperation(fmt.Sprintf("foo-%d-%d", id, i), make([]byte, 1000000)),
+				storage.SetOperation(fmt.Sprintf("bar-%d-%d", id, i), []byte("testValueBar")),
+			}
+			err := client.Batch(ctx, batchWrite...)
+			require.NoError(t, err)
+
+			err = client.Batch(ctx, storage.DeleteOperation(fmt.Sprintf("foo-%d-%d", id, i)))
+			require.NoError(t, err)
+
+			// Make sure the requests are somewhat spaced
+			time.Sleep(checkInterval * 2)
+
+			result, err := client.Get(ctx, fmt.Sprintf("foo-%d-%d", id, i))
+			require.NoError(t, err)
+			require.Equal(t, []byte(nil), result)
+
+			result, err = client.Get(ctx, fmt.Sprintf("bar-%d-%d", id, i))
+			require.NoError(t, err)
+			require.Equal(t, []byte("testValueBar"), result)
+
+			// Make sure the requests are somewhat spaced
+			time.Sleep(checkInterval)
+		}
+
+		wg.Done()
+	}
+
+	// Start a couple of concurrent threads and see how they add/remove data as needed without failures
+	for i := 0; i < 10; i++ {
+		wg.Add(1)
+		go clientOperationsThread(i)
+	}
+
+	wg.Wait()
```

Comment:
I think we should check that compactions actually happened in this test. Maybe pass in a zaptest/observer and assert on the logs?

Reply:
Great advice, thank you!